tools/wptrunner/wptrunner/testrunner.py

```diff
@@ -1020,18 +1038,32 @@ def wait(self, timeout: Optional[float] = None) -> None:
             timeout: Overall timeout (in seconds) for all threads to join. The
                 default value indicates an indefinite timeout.
         """
-        deadline = None if timeout is None else time.time() + timeout
-        for manager in self.pool:
-            manager_timeout = None
-            if deadline is not None:
-                manager_timeout = max(0, deadline - time.time())
-            manager.join(manager_timeout)
+        # Here, the main thread cannot simply `join()` the threads in
+        # `self.pool` sequentially because a keyboard interrupt raised during a
+        # `Thread.join()` may incorrectly mark that thread as "stopped" when it
+        # is not [0, 1]. Subsequent `join()`s for the affected thread won't
+        # block anymore, so a subsequent `ManagerGroup.wait()` may return with
+        # that thread still alive.
+        #
+        # To the extent the timeout allows, it's important that
+        # `ManagerGroup.wait()` returns with all `TestRunnerManager` threads
+        # actually stopped. Otherwise, a live thread may log after `mozlog`
+        # shutdown (not allowed) or worse, leak browser processes that the
+        # thread should have stopped when exiting its run loop [2].
+        #
+        # [0]: https://github.com/python/cpython/issues/90882
+        # [1]: https://github.com/python/cpython/blob/558b517b/Lib/threading.py#L1146-L1178
+        # [2]: https://crbug.com/330236796
+        assert self.stop_state, "ManagerGroup hasn't been started yet"
+        assert not self.stop_state.done.broken, self.stop_state
+        self.stop_state.done.wait(timeout)
```

Comment:
In what scenario we need this timeout value?

Reply:
But if the timeout does happen, that would mean the a `TestRunnerManager` is blocked somewhere and were not able to pass its Barrier? And that will be a leak?

I briefly checked the code. Looks like there is not anything that can block the TestRunnerManager, so the timeout is only for safety. I guess we can keep an eye on this to see if timeout can actually happen.

Reply:
> But if the timeout does happen, that would mean the a TestRunnerManager is blocked somewhere and were not able to pass its Barrier? And that will be a leak?

The answer to both questions is yes.

For the leak, I think marking `TestRunnerManager`s as `daemon=False` might be good to do anyway as a follow-up.

> Looks like there is not anything that can block the TestRunnerManager, so the timeout is only for safety.

Right, it should almost never be reached. Note that [stopping the browser](https://github.com/web-platform-tests/wpt/blob/d5ac8fb9826e3c5b25ad7521a207e0a25b057027/tools/wptrunner/wptrunner/testrunner.py#L242) can take arbitrarily long, depending on the vendor's implementation; the timeout guarantees a shutdown eventually in case of bad behavior.

Reply:
Looks like Firefox/Safari are using some sophisticated mechanism to ensure sessions are ended gracefully. And Chrome (actually webdriver browser) is just trying to kill the webdriver. Maybe we need check with chromedriver team to understand if chromedriver will kill the browser in this case. Or do we need to implement similar things in Wptrunner? 

[1] https://source.chromium.org/search?q=f:third_party%2Fwpt_tools%2Fwpt%2Ftools%2Fwptrunner%2Fwptrunner%2Fbrowsers%2F%20stop&ss=chromium%2Fchromium%2Fsrc

Reply:
I think `chromedriver`, not wptrunner, should be responsible for cleaning up `chrome` processes. That seems to match the actual behavior on `SIGTERM`:

```
→ xvfb-run chromedriver &
[1] 2677
→ Starting ChromeDriver 125.0.6381.0 (0000000000000000000000000000000000000000-0000000000000000000000000000000000000000) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
curl -s -d '{"capabilities": {}}' http://localhost:9515/session | jq
{
  "value": {
    "capabilities": {
      "acceptInsecureCerts": false,
      "browserName": "chrome",
      "browserVersion": "125.0.6381.0",
      "chrome": {
        "chromedriverVersion": "125.0.6381.0 (0000000000000000000000000000000000000000-0000000000000000000000000000000000000000)",
        "userDataDir": "/tmp/.org.chromium.Chromium.CPo9Od"
      },
      "fedcm:accounts": true,
      "goog:chromeOptions": {
        "debuggerAddress": "localhost:35529"
      },
      "networkConnectionEnabled": false,
      "pageLoadStrategy": "normal",
      "platformName": "linux",
      "proxy": {},
      "setWindowRect": true,
      "strictFileInteractability": false,
      "timeouts": {
        "implicit": 0,
        "pageLoad": 300000,
        "script": 30000
      },
      "unhandledPromptBehavior": "dismiss and notify",
      "webauthn:extension:credBlob": true,
      "webauthn:extension:largeBlob": true,
      "webauthn:extension:minPinLength": true,
      "webauthn:extension:prf": true,
      "webauthn:virtualAuthenticators": true
    },
    "sessionId": "32c900b778ed76f0bcbadbae697700e8"
  }
}
→ ps
    PID TTY          TIME CMD
   2560 pts/0    00:00:00 zsh
   2608 pts/0    00:00:00 zsh
   2677 pts/0    00:00:00 xvfb-run
   2686 pts/0    00:00:00 Xvfb
   2688 pts/0    00:00:02 chromedriver
   2719 pts/0    00:00:07 chrome
   2743 pts/0    00:00:00 chrome
   2744 pts/0    00:00:00 chrome
   2746 pts/0    00:00:00 chrome
   2772 pts/0    00:00:01 chrome
   2773 pts/0    00:00:00 chrome
   2823 pts/0    00:00:00 chrome
   2833 pts/0    00:00:00 chrome
   2847 pts/0    00:00:00 chrome
   2949 pts/0    00:00:00 chrome
   2973 pts/0    00:00:00 ps
→ kill 2688
→ Terminated

[1]  + 2677 exit 143   xvfb-run chromedriver
→ ps
    PID TTY          TIME CMD
   2560 pts/0    00:00:00 zsh
   2608 pts/0    00:00:00 zsh
   3340 pts/0    00:00:00 ps
```

CC @nechaev-chromium @sadym-chromium